src/neon/neon_radix4.rs

```diff
@@ -0,0 +1,489 @@
+    Len2(NeonF64Butterfly2<T>),
+    Len4(NeonF64Butterfly4<T>),
+    Len8(NeonF64Butterfly8<T>),
+    Len16(NeonF64Butterfly16<T>),
+    Len32(NeonF64Butterfly32<T>),
+}
+
+pub struct Neon32Radix4<T> {
+    _phantom: std::marker::PhantomData<T>,
+    twiddles: Box<[float32x4_t]>,
+
+    base_fft: Neon32Butterfly<T>,
+    base_len: usize,
+
+    len: usize,
+    direction: FftDirection,
+    bf4: NeonF32Butterfly4<T>,
+
+    shuffle_map: Box<[usize]>,
+}
+
+impl<T: FftNum> Neon32Radix4<T> {
+    /// Preallocates necessary arrays and precomputes necessary data to efficiently compute the power-of-two FFT
+    pub fn new(len: usize, direction: FftDirection) -> Self {
+        assert!(
+            len.is_power_of_two(),
+            "Radix4 algorithm requires a power-of-two input size. Got {}",
+            len
+        );
+        assert_f32::<T>();
+
+        // figure out which base length we're going to use
+        let num_bits = len.trailing_zeros();
+        let (base_len, base_fft) = match num_bits {
+            0 => (
+                len,
+                Neon32Butterfly::Len1(NeonF32Butterfly1::new(direction)),
+            ),
+            1 => (
+                len,
+                Neon32Butterfly::Len2(NeonF32Butterfly2::new(direction)),
+            ),
+            2 => (
+                len,
+                Neon32Butterfly::Len4(NeonF32Butterfly4::new(direction)),
+            ),
+            3 => (
+                len,
+                Neon32Butterfly::Len8(NeonF32Butterfly8::new(direction)),
+            ),
+            _ => {
+                if num_bits % 2 == 1 {
+                    if len < USE_BUTTERFLY32_FROM {
+                        (8, Neon32Butterfly::Len8(NeonF32Butterfly8::new(direction)))
+                    } else {
+                        (
+                            32,
+                            Neon32Butterfly::Len32(NeonF32Butterfly32::new(direction)),
+                        )
+                    }
+                } else {
+                    (
+                        16,
+                        Neon32Butterfly::Len16(NeonF32Butterfly16::new(direction)),
+                    )
+                }
+            }
+        };
+
+        // precompute the twiddle factors this algorithm will use.
+        // we're doing the same precomputation of twiddle factors as the mixed radix algorithm where width=4 and height=len/4
+        // but mixed radix only does one step and then calls itself recusrively, and this algorithm does every layer all the way down
+        // so we're going to pack all the "layers" of twiddle factors into a single array, starting with the bottom layer and going up
+        let mut twiddle_stride = len / (base_len * 4);
+        let mut twiddle_factors = Vec::with_capacity(len * 2);
+        while twiddle_stride > 0 {
+            let num_rows = len / (twiddle_stride * 4);
+            for i in 0..num_rows / 2 {
+                for k in 1..4 {
+                    unsafe {
```

Comment:
If the only reason we need this unsafe tag is for line 131, could we move the unsafe tag to line 131?